### PR TITLE
Zeotap ID+ submodule

### PR DESCRIPTION
**Pull request by shikharsharma-zeotap**
## Type of change
- [x] Feature

## Description of change
This pr adds the **zeotapIdPlusIdSystem** submodule to add Zeotap **IDP** id to the bid requests.

Documentation PR Link: [https://github.com/prebid/prebid.github.io/pull/2260](https://github.com/prebid/prebid.github.io/pull/2260)

### CODE REVIEW

File: integrationExamples/gpt/userId_example.html
Comment (smenzer) [inline integrationExamples/gpt/userId_example.html:224]: it looks like you are defining your own name for the cookie and you're not leveraging the userId module storage (which is what this bit is used for). so I think you can just ignore this, like how Parrable does.

based on your module code, I think you can just do this:

```suggestion
                    },
                    {
                      name: "zeotapIdPlus"
```

File: modules/zeotapIdPlusIdSystem.js
Comment (smenzer) [inline modules/zeotapIdPlusIdSystem.js:24]: in your example you don't list any params, but you are requiring "name" here. what is the name param and how are you using it? i don't see it anywhere in your code

File: modules/zeotapIdPlusIdSystem.js
Comment (smenzer) [inline modules/zeotapIdPlusIdSystem.js:35]: per the above, i don't think you need this since you're not using any `configParams`
```suggestion
```

File: test/spec/modules/userId_spec.js
Comment (smenzer) [inline test/spec/modules/userId_spec.js:441]: per above, this can probably be dropped
```suggestion
            name: 'zeotapIdPlus'
```

File: test/spec/modules/userId_spec.js
Comment (smenzer) [inline test/spec/modules/userId_spec.js:1361]: per above, this can probably just be your name
```suggestion
            name: 'zeotapIdPlus'
```

**Review by smenzer — CHANGES_REQUESTED** [review]
a few changes requested based on not seeing the need for storage configuration or configParams mostly

File: integrationExamples/gpt/userId_example.html
Comment (shikharsharma-zeotap) [reply integrationExamples/gpt/userId_example.html]: Hey @smenzer, thanks for the suggestion, have made changes to the module and files to remove storage configuration.

File: modules/zeotapIdPlusIdSystem.js
Comment (shikharsharma-zeotap) [reply modules/zeotapIdPlusIdSystem.js]: This is removed, not required.

File: modules/zeotapIdPlusIdSystem.js
Comment (shikharsharma-zeotap) [reply modules/zeotapIdPlusIdSystem.js]: made changes.

File: test/spec/modules/userId_spec.js
Comment (shikharsharma-zeotap) [reply test/spec/modules/userId_spec.js]: dropped.

File: test/spec/modules/userId_spec.js
Comment (shikharsharma-zeotap) [reply test/spec/modules/userId_spec.js]: made changes.

File: test/spec/modules/zeotapIdPlusIdSystem_spec.js
Comment (shikharsharma-zeotap) [reply test/spec/modules/zeotapIdPlusIdSystem_spec.js]: dropped.

File: test/spec/modules/zeotapIdPlusIdSystem_spec.js
Comment (shikharsharma-zeotap) [reply test/spec/modules/zeotapIdPlusIdSystem_spec.js]: dropped.

File: test/spec/modules/zeotapIdPlusIdSystem_spec.js
Comment (shikharsharma-zeotap) [reply test/spec/modules/zeotapIdPlusIdSystem_spec.js:67]: Thanks for pointing this out @smenzer , the purpose of this test was to check whether zeotap id is returned if id is present in local storage instead of cookies.
Have made changes to the test.

**Comment by shikharsharma-zeotap** [comment]
@smenzer anything else required from my side for this to be merged?

**Comment by smenzer** [comment]
@shikharsharma-zeotap where are you actually generating the ID? if there's no ID in cookie/local storage...right now you just return `undefined`...is that what you want? No calls to your servers are required?

**Comment by shikharsharma-zeotap** [comment]
@smenzer the ID will be generated on the publisher page by our JS. No calls to our server is required from prebid's end.

**Comment by patmmccann** [comment]
Are modifications to the eids spec and userid spec required parts of adding a user id module? Most user id commits seem to modify them but not all.

**Comment by smenzer** [comment]
> Are modifications to the eids spec and userid spec required parts of adding a user id module? Most user id commits seem to modify them but not all. You seem to be missing a unit test on eids spec

there are a couple tests in those two specs that look at basic testing for any modules that want to. eids seems required although the tests in userId_spec can be covered by a separate test suite; but better to have both than nothing if you ask me.

**Comment by smenzer** [comment]
> @smenzer the ID will be generated on the publisher page by our JS. No calls to our server is required from prebid's end.

so if the publisher doesn't have your JS on page from a direct integration with Zeotap, you aren't expecting to be able to provide an ID, right?

**Comment by shikharsharma-zeotap** [comment]
Yes @smenzer, we expect publisher integration with zeotap for ID to be present, so generation/server calls from prebid are not required. Also, thanks @patmmccann , will be adding the Eid spec tests too.

**Comment by smenzer** [comment]
@idettman can you do a quick review and make sure you're ok with this before we merge? thanks!

**Comment by patmmccann** [comment]
> @idettman can you do a quick review and make sure you're ok with this before we merge? thanks!

I don't think this can be merged until eids.js is covered with a test

File: modules/zeotapIdPlusIdSystem.js
Comment (patmmccann) [inline modules/zeotapIdPlusIdSystem.js:11]: should this be configurable by the pub? can you read if this is set in config and fallback to 'IDP' ?

Comment (shikharsharma-zeotap) [reply modules/zeotapIdPlusIdSystem.js:11]: @patmmccann I have added the tests to eid spec file. The cookie name 'IDP' is not configurable by the pub. It gets added to the storage by Zeotap JS after publisher integration. I didn't get the other question.

Comment (patmmccann) [reply modules/zeotapIdPlusIdSystem.js:11]: Thanks!

**Comment by shikharsharma-zeotap** [comment]
@idettman waiting for your comments. @smenzer @patmmccann have resolved conflicts with base branch. Can we merge this now?

**Review by smenzer — COMMENTED** [review]
thanks for the additional test for eids. 

CI is failing, though, so that needs to be fixed before we can merge. @idettman can you do a quick review, too, please?

**Comment by shikharsharma-zeotap** [comment]
@smenzer the circleCI shows 2 tests breaking during the BrowserStack testing.
One is related to Parrable Id Module, other is for Zeotap Id module. Is there something that I require to fix from my end, so that CI looks green?

**Comment by smenzer** [comment]
@shikharsharma-zeotap i'm not really sure why the builds are failing...locally they run fine for me. i've reached out to the prebid community in slack and perhaps someone else here can take a look at things. once that is fixed this is good to go from my end

**Comment by smenzer** [comment]
@shikharsharma-zeotap can you try pulling in the latest master to see if that helps at all?

**Comment by shikharsharma-zeotap** [comment]
Hey @smenzer , have rebased with the latest master, looks like the circleCI has stopped complaining for us now. Thanks!

**Comment by smenzer** [comment]
> Hey @smenzer , have rebased with the latest master, looks like the circleCI has stopped complaining for us now. Thanks!

great! I'd like to get a review from @jlukas79 and then this can be merged.

**Review by jlukas79 — APPROVED** [review]

**Pull request merged**
